Code/DataStructs/MultiFPBReader.cpp

```diff
@@ -0,0 +1,293 @@
+    if (initOnSearch) (*readers)[i]->init();
+    (*accum)[i] = (*readers)[i]->getContainingNeighbors(bv);
+  }
+}
+
+void get_containing_nbrs(
+    const std::vector<FPBReader *> &d_readers, const boost::uint8_t *bv,
+    std::vector<std::pair<unsigned int, unsigned int> > &res,
+    unsigned int numThreads, bool initOnSearch) {
+  numThreads = getNumThreadsToUse(numThreads);
+#ifdef RDK_THREADSAFE_SSS
+  boost::thread_group tg;
+#endif
+
+  std::vector<std::vector<unsigned int> > accum(d_readers.size());
+  if (numThreads == 1) {
+    contain_helper(0, 1, bv, &d_readers, &accum, initOnSearch);
+  }
+#ifdef RDK_THREADSAFE_SSS
+  else {
+    for (unsigned int tid = 0; tid < numThreads && tid < d_readers.size();
+         ++tid) {
+      tg.add_thread(new boost::thread(contain_helper, tid, numThreads, bv,
+                                      &d_readers, &accum, initOnSearch));
+    }
+    tg.join_all();
+  }
+#endif
+
+  res.clear();
+  for (unsigned int i = 0; i < d_readers.size(); ++i) {
+    std::vector<unsigned int> &r_res = accum[i];
+    BOOST_FOREACH (unsigned int ri, r_res) {
+      res.push_back(std::make_pair(ri, i));
+    }
+  }
+
+  std::sort(res.begin(), res.end(), pairSorter());
+}
+
+}  // end of anonymous namespace
+
+void MultiFPBReader::init() {
+  unsigned int nBits = 0;
+  BOOST_FOREACH (FPBReader *rdr, d_readers) {
+    rdr->init();
+    if (!nBits) {
+      nBits = rdr->nBits();
+    } else {
+      if (rdr->nBits() != nBits)
+        throw ValueErrorException("bit lengths of child readers don't match");
+    }
+  }
+  df_init = true;
+};
+
+MultiFPBReader::MultiFPBReader(std::vector<FPBReader *> &readers,
+                               bool takeOwnership, bool initOnSearch) {
+  df_init = false;
+  df_takeOwnership = takeOwnership;
+  df_initOnSearch = initOnSearch;
+  BOOST_FOREACH (FPBReader *rdr, readers) {
+    PRECONDITION(rdr != NULL, "bad reader");
+  }
+  d_readers = readers;
+}
+
+FPBReader *MultiFPBReader::getReader(unsigned int which) {
+  URANGE_CHECK(which, d_readers.size());
+  return d_readers[which];
+}
+unsigned int MultiFPBReader::nBits() const {
+  PRECONDITION(d_readers.size(), "no readers");
+  PRECONDITION(df_init, "not initialized");
+  return d_readers[0]->nBits();
+}
+
+std::vector<MultiFPBReader::ResultTuple> MultiFPBReader::getTanimotoNeighbors(
+    const boost::uint8_t *bv, double threshold, int numThreads) const {
+  PRECONDITION(df_init || df_initOnSearch, "not initialized");
+
+  std::vector<MultiFPBReader::ResultTuple> res;
+  get_tani_nbrs(d_readers, bv, threshold, res, numThreads, df_initOnSearch);
+  return res;
+}
+
+std::vector<MultiFPBReader::ResultTuple> MultiFPBReader::getTanimotoNeighbors(
+    const ExplicitBitVect &ebv, double threshold, int numThreads) const {
+  PRECONDITION(df_init || df_initOnSearch, "not initialized");
+  std::vector<MultiFPBReader::ResultTuple> res;
+  boost::uint8_t *bv = detail::bitsetToBytes(*(ebv.dp_bits));
```

Comment:
Yeah, I think it's fine at the moment. When we make the modern C++ switch this is one of the (many) places that can be cleaned up.